File: leetcode/97.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        if(s3.size() != s1.size() + s2.size()) return false;
        vector<vector<bool>> dp(s1.size()+1, vector<bool>(s2.size()+1, false));

        for(int i = 0; i <= s1.size(); ++i) {
            for(int j = 0; j <= s2.size(); ++j) {
                // deal with dp[i][j]
                if(i == 0 && j == 0) {
                    dp[i][j] = true;
                }
                else if(i == 0) {
                    dp[i][j] = dp[i][j-1] && s3[i+j-1] == s2[j-1];
                }
                else if(j == 0) {
                    dp[i][j] = dp[i-1][j] && s3[i+j-1] == s1[i-1];
                }
                else {
                    dp[i][j] = dp[i][j] || (s3[i+j-1] == s1[i-1] && dp[i-1][j]);
                    dp[i][j] = dp[i][j] || (s3[i+j-1] == s2[j-1] && dp[i][j-1]);
                }
            }
        }
        return dp[s1.size()][s2.size()];
    }
};

File: leetcode/97.py
class Solution:

    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        self.dp = {}
        return self.isInterleaveInternal(s1, 0, s2, 0, s3, 0)

    def isInterleaveInternal(self, s1: str, i1: int, s2: str, i2: int, s3: str, i3: int) -> bool:
        if i3 == len(s3):
            return i1 == len(s1) and i2 == len(s2) and i3 == len(s3)

        cache_key = (i1, i2, i3)
        dp = self.dp
        if cache_key in dp:
            return dp[cache_key]
        dp[cache_key] = False

        if i1 < len(s1) and s1[i1] == s3[i3]:
            dp[cache_key] = dp[cache_key] or self.isInterleaveInternal(
                s1, i1+1, s2, i2, s3, i3+1)

        if i2 < len(s2) and s2[i2] == s3[i3]:
            dp[cache_key] = dp[cache_key] or self.isInterleaveInternal(
                s1, i1, s2, i2+1, s3, i3+1)

        return dp[cache_key]
